src/contact.rs

```diff
@@ -189,6 +190,128 @@ pub async fn make_vcard(context: &Context, contacts: &[ContactId]) -> Result<Str
     Ok(contact_tools::make_vcard(&vcard_contacts))
 }
 
+/// Imports contacts from the given vCard.
+///
+/// Returns the ids of created/modified contacts in the order they appear in `vcard`.
+pub async fn import_vcard(context: &Context, vcard: &str) -> Result<Vec<ContactId>> {
+    let contacts = contact_tools::parse_vcard(vcard);
+    let mut contact_ids = Vec::with_capacity(contacts.len());
+    for c in &contacts {
+        let Ok(Some(id)) = import_vcard_contact(context, c)
+            .await
+            .with_context(|| format!("import_vcard_contact() failed for {}", c.addr))
+            .log_err(context)
+        else {
+            continue;
+        };
+        contact_ids.push(id);
+    }
+    Ok(contact_ids)
+}
+
+async fn import_vcard_contact(
+    context: &Context,
+    contact: &VcardContact,
+) -> Result<Option<ContactId>> {
+    let (id, modified) = match ContactAddress::new(&contact.addr) {
+        Err(e) => return Err(e).context("Invalid address"),
+        Ok(addr) => {
+            // Maybe we haven't yet send messages to this contact, but importing the vCard is also
+            // an explicit user action, so it's ok. We mustn't use `Origin::AddressBook` here
+            // because the vCard may be created not by us, also we want `contact.authname` to be
+            // saved as the authname and not a locally given name.
+            let origin = Origin::OutgoingTo;
+            match Contact::add_or_lookup(context, &contact.authname, &addr, origin).await {
+                Err(e) => return Err(e).context("Contact::add_or_lookup() failed"),
+                Ok((ContactId::SELF, _)) => return Ok(None),
+                Ok(val) => val,
+            }
+        }
+    };
+    if modified != Modifier::None {
+        context.emit_event(EventType::ContactsChanged(Some(id)));
+    }
+    let key = contact.key.as_ref().and_then(|k| {
+        SignedPublicKey::from_base64(k)
+            .with_context(|| {
+                format!(
+                    "import_vcard_contact: Cannot decode key for {}",
+                    contact.addr
+                )
+            })
+            .log_err(context)
+            .ok()
+    });
+    if let Some(public_key) = key {
+        let timestamp = *contact.timestamp.as_ref().unwrap_or(&time());
+        let aheader = Aheader {
+            addr: contact.addr.clone(),
+            public_key,
+            prefer_encrypt: EncryptPreference::NoPreference,
```

Comment:
This should be

```suggestion
            prefer_encrypt: EncryptPreference::Mutual,
```

in order to make it work with chatmail. Plus, after we just imported a vCard with a public key, I think that we can let this count as "this contact would like to encrypt". If we were wrong, they will probably answer with "I can't read your message" and the encryption preference will be reset.